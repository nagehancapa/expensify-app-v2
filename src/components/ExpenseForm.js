import React, { useState } from "react";
import DatePicker from "react-date-picker";
import "react-date-picker/dist/DatePicker.css";
import "react-calendar/dist/Calendar.css";

const ExpenseForm = () => {
  const [description, setDescription] = useState("");
  const [note, setNote] = useState("");
  const [amount, setAmount] = useState("");
  const [createdAt, setCreatedAt] = useState(new Date());

  return (
    <div>
      <form>
        <input
          type="text"
          placeholder="Description"
          autoFocus
          value={description}
          onChange={(e) => setDescription(e.target.value)}
        />
        <input
          type="text"
          placeholder="Amount"
          value={amount}
          onChangeText={(e) => {
            //let userInput = amount;
            //if (!amount || amount.match(/^[0-9]{1,2}([.][0-9]{1,2})?$/)) {
            //(/^\d{1,}(\.\d{0,2})?$/)) {
            setAmount(e.target.value);
            //} else {
            // console.log("wtf");
            //}
          }}
        />
        <DatePicker value={createdAt} onChange={(date) => setCreatedAt(date)} />
        <textarea
          placeholder="Add a note for your expense (optional)"
          value={note}
          onChange={(e) => setNote(e.target.value)}
        ></textarea>
        <button>Add Expense</button>
      </form>
    </div>
  );
};

export default ExpenseForm;
