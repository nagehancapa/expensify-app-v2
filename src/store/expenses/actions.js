import { push, ref, set } from "firebase/database";
import db from "../../firebase/firebase";

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;
    const expense = {
      description,
      note,
      amount,
      createdAt,
    };
    const newRef = push(ref(db, "expenses"));
    return set(newRef, expense)
      .then(() => {
        // dispatch to change redux store
        dispatch(
          addExpense({
            id: newRef.key,
            ...expense,
          })
        );
      })
      .catch((e) => {
        console.log("This failed.", e);
      });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id,
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});
