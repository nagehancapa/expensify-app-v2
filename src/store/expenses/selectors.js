// Get visible expenses
// destructure filters
// const getVisibleExpenses = (expenses, filters)
export const getVisibleExpenses = (
  expenses,
  { text, sortBy, startDate, endDate }
) => {
  return expenses
    .filter((expense) => {
      const startDateMatch =
        typeof startDate !== "number" || expense.createdAt >= startDate;
      const endDateMatch =
        typeof endDate !== "number" || expense.createdAt <= endDate;
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        // return -1 if a come first, 1 if b come first
        // most recent expense up top
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === "amount") {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};

export const selectExpenses = (state) => state.expenses;
