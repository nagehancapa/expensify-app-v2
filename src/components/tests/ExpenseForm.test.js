import React from "react";
import { shallow } from "enzyme";
import ExpenseForm from "../ExpenseForm";
import expenses from "../../fixtures/expenses";
// import toJson from "enzyme-to-json";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "jest-canvas-mock";

beforeAll(() => {
  jest.useFakeTimers("modern");
  jest.setSystemTime(new Date("January 1, 1970"));
});

afterAll(() => {
  jest.useRealTimers();
});

test("should render ExpenseForm correctly", () => {
  const wrapper = shallow(<ExpenseForm />);
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseForm correctly with expense data", () => {
  const wrapper = shallow(<ExpenseForm expense={expenses[1]} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render error for invalid form submission", () => {
  const wrapper = shallow(<ExpenseForm />);
  expect(wrapper).toMatchSnapshot();
  wrapper.find("form").simulate("submit", {
    preventDefault: () => {},
  });
  // console.log(toJson(wrapper.find("p")));
  const error = wrapper.find("p").children();
  expect(error.isEmptyRender()).toBe(false);
  expect(wrapper).toMatchSnapshot();
});

test("should set description on input change", () => {
  const value = "New description";
  const wrapper = shallow(<ExpenseForm />);
  //at(0) means first input
  wrapper.find("input").at(0).simulate("change", {
    target: { value },
  });
  // console.log(wrapper.find("input").get(0).props.value);
  expect(wrapper.find("input").get(0).props.value).toBe(value);
});

test("should set description on input change test with testing library", () => {
  render(<ExpenseForm />);
  const input = screen.getByPlaceholderText("Description");
  userEvent.type(input, "second description");
  expect(screen.getByPlaceholderText("Description")).toHaveValue(
    "second description"
  );
});

test("should set note on textarea change", () => {
  const value = "New note";
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find("textarea").simulate("change", {
    target: { value },
  });
  // console.log(wrapper.find("textarea").get(0).props.value);
  expect(wrapper.find("textarea").get(0).props.value).toBe(value);
});

test("should set note on textare change test with testing library", () => {
  render(<ExpenseForm />);
  const textarea = screen.getByPlaceholderText(/Add a note/);
  userEvent.type(textarea, "second note");
  expect(screen.getByPlaceholderText(/Add a note/)).toHaveValue("second note");
});
