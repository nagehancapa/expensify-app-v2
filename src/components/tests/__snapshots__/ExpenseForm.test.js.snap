// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseForm correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Description"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Amount"
    type="text"
    value=""
  />
  <DatePicker
    calendarIcon={
      <svg
        className="react-date-picker__calendar-button__icon react-date-picker__button__icon"
        height={19}
        stroke="black"
        strokeWidth={2}
        viewBox="0 0 19 19"
        width={19}
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="15"
          width="15"
          x="2"
          y="2"
        />
        <line
          x1="6"
          x2="6"
          y1="0"
          y2="4"
        />
        <line
          x1="13"
          x2="13"
          y1="0"
          y2="4"
        />
      </svg>
    }
    clearIcon={
      <svg
        className="react-date-picker__clear-button__icon react-date-picker__button__icon"
        height={19}
        stroke="black"
        strokeWidth={2}
        viewBox="0 0 19 19"
        width={19}
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="4"
          x2="15"
          y1="4"
          y2="15"
        />
        <line
          x1="15"
          x2="4"
          y1="4"
          y2="15"
        />
      </svg>
    }
    closeCalendar={true}
    isOpen={null}
    onChange={[Function]}
    openCalendarOnFocus={true}
    returnValue="start"
    value={1969-12-31T23:00:00.000Z}
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a note for your expense (optional)"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`should render ExpenseForm correctly with expense data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Description"
    type="text"
    value="Rent"
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Amount"
    type="text"
    value="1095"
  />
  <DatePicker
    calendarIcon={
      <svg
        className="react-date-picker__calendar-button__icon react-date-picker__button__icon"
        height={19}
        stroke="black"
        strokeWidth={2}
        viewBox="0 0 19 19"
        width={19}
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="15"
          width="15"
          x="2"
          y="2"
        />
        <line
          x1="6"
          x2="6"
          y1="0"
          y2="4"
        />
        <line
          x1="13"
          x2="13"
          y1="0"
          y2="4"
        />
      </svg>
    }
    clearIcon={
      <svg
        className="react-date-picker__clear-button__icon react-date-picker__button__icon"
        height={19}
        stroke="black"
        strokeWidth={2}
        viewBox="0 0 19 19"
        width={19}
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="4"
          x2="15"
          y1="4"
          y2="15"
        />
        <line
          x1="15"
          x2="4"
          y1="4"
          y2="15"
        />
      </svg>
    }
    closeCalendar={true}
    isOpen={null}
    onChange={[Function]}
    openCalendarOnFocus={true}
    returnValue="start"
    value={1969-12-27T23:00:00.000Z}
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a note for your expense (optional)"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`should render error for invalid form submission 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Description"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Amount"
    type="text"
    value=""
  />
  <DatePicker
    calendarIcon={
      <svg
        className="react-date-picker__calendar-button__icon react-date-picker__button__icon"
        height={19}
        stroke="black"
        strokeWidth={2}
        viewBox="0 0 19 19"
        width={19}
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="15"
          width="15"
          x="2"
          y="2"
        />
        <line
          x1="6"
          x2="6"
          y1="0"
          y2="4"
        />
        <line
          x1="13"
          x2="13"
          y1="0"
          y2="4"
        />
      </svg>
    }
    clearIcon={
      <svg
        className="react-date-picker__clear-button__icon react-date-picker__button__icon"
        height={19}
        stroke="black"
        strokeWidth={2}
        viewBox="0 0 19 19"
        width={19}
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="4"
          x2="15"
          y1="4"
          y2="15"
        />
        <line
          x1="15"
          x2="4"
          y1="4"
          y2="15"
        />
      </svg>
    }
    closeCalendar={true}
    isOpen={null}
    onChange={[Function]}
    openCalendarOnFocus={true}
    returnValue="start"
    value={1969-12-31T23:00:00.000Z}
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a note for your expense (optional)"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`should render error for invalid form submission 2`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="form__error"
  >
    Please provide description and amount.
  </p>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Description"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Amount"
    type="text"
    value=""
  />
  <DatePicker
    calendarIcon={
      <svg
        className="react-date-picker__calendar-button__icon react-date-picker__button__icon"
        height={19}
        stroke="black"
        strokeWidth={2}
        viewBox="0 0 19 19"
        width={19}
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="15"
          width="15"
          x="2"
          y="2"
        />
        <line
          x1="6"
          x2="6"
          y1="0"
          y2="4"
        />
        <line
          x1="13"
          x2="13"
          y1="0"
          y2="4"
        />
      </svg>
    }
    clearIcon={
      <svg
        className="react-date-picker__clear-button__icon react-date-picker__button__icon"
        height={19}
        stroke="black"
        strokeWidth={2}
        viewBox="0 0 19 19"
        width={19}
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="4"
          x2="15"
          y1="4"
          y2="15"
        />
        <line
          x1="15"
          x2="4"
          y1="4"
          y2="15"
        />
      </svg>
    }
    closeCalendar={true}
    isOpen={null}
    onChange={[Function]}
    openCalendarOnFocus={true}
    returnValue="start"
    value={1969-12-31T23:00:00.000Z}
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a note for your expense (optional)"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save Expense
    </button>
  </div>
</form>
`;
